client/session.go

```diff
@@ -0,0 +1,277 @@
+			if routeErr != nil {
+				return routeErr
+			}
+			if !shardHasMinConnectionsToOneHost {
+				clusterHasMinConnectionsToAllShards = false
+				break
+			}
+		}
+		if clusterHasMinConnectionsToAllShards {
+			// All done
+			break
+		}
+		time.Sleep(clusterConnectWaitInterval)
+	}
+
+	s.Lock()
+	prev := s.queues
+	s.queues = queues
+	s.setWriteOpPoolsWithLock(topologyMap)
+	s.topoMap = topologyMap
+	s.Unlock()
+
+	// Asynchronously close the previous set of host queues
+	go func() {
+		for i := range prev {
+			prev[i].Close()
+		}
+	}()
+
+	return nil
+}
+
+func (s *session) newHostQueue(host m3db.Host, topologyMap m3db.TopologyMap) hostQueue {
+	// NB(r): Due to hosts being replicas we have:
+	// = replica * numWrites
+	// = total writes to all hosts
+	// We need to pool:
+	// = replica * (numWrites / writeBatchSize)
+	// = number of batch request structs to pool
+	totalBatches := topologyMap.Replicas() *
+		int(math.Ceil(float64(s.opts.GetWriteOpPoolSize())/float64(s.opts.GetWriteBatchSize())))
+	hostBatches := int(math.Ceil(float64(totalBatches) / float64(topologyMap.HostsLen())))
+	writeBatchRequestPool := newWriteBatchRequestPool(hostBatches)
+	writeRequestArrayPool := newWriteRequestArrayPool(hostBatches, s.opts.GetWriteBatchSize())
+	return newHostQueue(host, writeBatchRequestPool, writeRequestArrayPool, s.opts)
+}
+
+func (s *session) setWriteOpPoolsWithLock(topologyMap m3db.TopologyMap) {
+	totalShards := len(topologyMap.ShardScheme().All().Shards())
+	if len(s.writeOpPools) == totalShards {
+		// Already created all pools
+		return
+	}
+
+	s.writeOpPools = make([]writeOpPool, totalShards)
+	totalSize := s.opts.GetWriteOpPoolSize()
+	eachSize := totalSize / totalShards
+	for i := range s.writeOpPools {
+		s.writeOpPools[i] = newWriteOpPool(eachSize)
+	}
+}
+
+func (s *session) Write(id string, t time.Time, value float64, unit xtime.Unit, annotation []byte) error {
+	var (
+		w             *writeOp
+		pool          writeOpPool
+		wg            sync.WaitGroup
+		enqueueErr    error
+		enqueued      int
+		resultErrLock sync.Mutex
+		resultErr     error
+		resultErrs    int
+		quorum        int
+	)
+
+	timeType, timeTypeErr := convert.UnitToTimeType(unit)
+	if timeTypeErr != nil {
+		return timeTypeErr
+	}
+
+	ts, tsErr := convert.TimeToValue(t, timeType)
+	if tsErr != nil {
+		return tsErr
+	}
+
+	s.RLock()
+	routeErr := s.topoMap.RouteForEach(id, func(shard uint32) {
+		pool = s.writeOpPools[shard]
+		w = pool.Get()
+		w.request.ID = id
+		w.datapoint.Value = value
+		w.datapoint.Timestamp = ts
+		w.datapoint.TimestampType = timeType
+		w.datapoint.Annotation = annotation
+		w.completionFn = func(result interface{}, err error) {
+			if err != nil {
+				resultErrLock.Lock()
+				if resultErr == nil {
+					resultErr = err
+				}
+				resultErrs++
+				resultErrLock.Unlock()
+			}
+			wg.Done()
+		}
+	}, func(idx int, host m3db.Host) {
+		wg.Add(1)
+		enqueued++
+		if err := s.queues[idx].Enqueue(w); err != nil && enqueueErr != nil {
```

Comment:
Did you mean `enqueueErr == nil`? otherwise `enqueueErr` will never become non-nil